src/persist/src/operators/source.rs

```diff
@@ -0,0 +1,184 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! A Timely Dataflow operator that mirrors a persisted stream.
+
+use std::sync::mpsc::{self, TryRecvError};
+
+use timely::dataflow::operators::generic::operator;
+use timely::dataflow::operators::{Concat, ToStream};
+use timely::dataflow::{Scope, Stream};
+use timely::Data;
+
+use crate::indexed::runtime::{StreamReadHandle, StreamWriteHandle};
+use crate::indexed::TailEvent;
+use crate::{operators, Token};
+
+/// A Timely Dataflow operator that mirrors a persisted stream.
+pub trait PersistedSource<G: Scope<Timestamp = u64>, K: Data, V: Data> {
+    /// Emits a snapshot of the persisted stream taken as of this call and tails
+    /// any new data added to the persisted stream after that.
+    fn persisted_source(
+        &mut self,
+        read: Token<StreamWriteHandle<K, V>, StreamReadHandle<K, V>>,
+    ) -> (
+        Stream<G, ((K, V), u64, isize)>,
+        Stream<G, (String, u64, isize)>,
+    );
+}
+
+impl<G, K, V> PersistedSource<G, K, V> for G
+where
+    G: Scope<Timestamp = u64>,
+    K: Data + Send,
+    V: Data + Send,
+{
+    fn persisted_source(
+        &mut self,
+        token: Token<StreamWriteHandle<K, V>, StreamReadHandle<K, V>>,
+    ) -> (
+        Stream<G, ((K, V), u64, isize)>,
+        Stream<G, (String, u64, isize)>,
+    ) {
+        let (_, read) = token.into_inner();
+        let (tail_tx, tail_rx) = mpsc::channel();
+        let tail_fn = Box::new(move |e| {
+            // TODO: If send fails, it means the operator is no longer running.
+            // We should probably allow the tail to deregister itself.
+            let _ = tail_tx.send(e);
+        });
+
+        let err_new = match read.tail(tail_fn) {
+            Ok(_) => operator::empty(self),
+            Err(err) => vec![(err.to_string(), 0, 1)].to_stream(self),
+        };
+
+        // TODO: Plumb the name of the stream down through the handles and use
+        // it for the operator.
+        let ok_new = operator::source(self, "PersistedSource", |capability, info| {
+            // TODO: Hook the activator up to the callback instead of
+            // TryRecvError::Empty.
+            let activator = self.activator_for(&info.address[..]);
+            let mut cap = Some(capability);
+            move |output| {
+                let mut done = false;
+                if let Some(cap) = cap.as_mut() {
+                    let mut session = output.session(cap);
+                    match tail_rx.try_recv() {
+                        Ok(e) => match e {
+                            TailEvent::Records(mut records) => {
+                                session.give_vec(&mut records);
```

Comment:
it took me a while to reason about what happens with multiple workers but I think:

each worker will register its own individual tail with their own mpsc queue and then any sent data will get broadcast over all of those queues to all of the workers, and we will end up with N copies of the data iff there are N workers is that right? this seems we want to only do all of this on worker 0 or something like that for the v1?

Reply:
Done. Added an ignored test. There's a decent bit of work to do to fix it up, but I'd prefer to do it in a followup, if that's alright.